### PR TITLE
Implement projection and predicate pushdown for CTEs.

### PR DESCRIPTION
**Pull request by mlyublena**
Example:
  Before Transformation:
```
    CTEProducer(cteX)
    |-- SomeOp
    `--Filter (Pred1)
       -- Projection (C1,C2)
          -- CTEConsumer(cteX)
    |-- ...
    `--Filter (Pred2)
       -- Projection (C3,C4)
          -- CTEConsumer(cteX)
 ```

  After Transformation:
```
    CTEProducer(cteX)
    |-- Filter (Pred1 or Pred2)
        -- Projection (C1,C2,C3,C4)
           -- SomeOp
    `--Filter (Pred1)
       -- Projection (C1,C2)
          -- CTEConsumer(cteX)
    |-- ...
    `--Filter (Pred2)
       -- Projection (C3,C4)
          -- CTEConsumer(cteX)
```

Note: will add more plan tests later.

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by jaystarshot — COMMENTED** [review]
Thanks for the fix, great code. Took a first glance, will try to complete the review as fast as I can.
Is it better to add filter and project pushdown rules in this optimizer itself after the rewrite so that filters and projects are pushed further into the producer tree?

**Comment by jaystarshot** [comment]
Nvm mind   I think there are other optimizers after this one which do the pushdowns

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java [hunk 1]
```diff
@@ -793,6 +793,35 @@ public void testComplexPersistentCteForViewQueries()
                 "DROP View non_persistent_view");
     }
 
+    public void testCteProjectionPushDown()
```
Comment (jaystarshot) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:796]: This test doesn't really check if the pushdowns happened, I would suggest using https://github.com/prestodb/presto/blob/3f2ed80ca1a9f464804ad6ae2cfabff3cfd5c77c/presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestLogicalCteOptimizer.java#L36

**Comment by mlyublena** [comment]
> Thanks for the fix, great code. Took a first glance, will try to complete the review as fast as I can. Is it better to add filter and project pushdown rules in this optimizer itself after the rewrite so that filters and projects are pushed further into the producer tree?

great suggestion! I'll add this

**Comment by jaystarshot** [comment]
Ahh actually it might not be needed since there are other optimizers after this optimizer one which do the pushdowns. [link](https://github.com/prestodb/presto/blob/8ce9c81e4c28ce5d2ffd9722c68867e889f81fe4/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java#L803)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java (hunk 1 above)
Comment (mlyublena) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:795]: Added a plan match test similar to the one above here:
presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestCteProjectionAndPredicatePushdown.java

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+            List<VariableReferenceExpression> columns = node.getOutputVariables();
+            context.addCteProducerInfo(cteName, columns);
+
+            return super.visitCteProducer(node, context);
+        }
+
+        public Void visitFilter(FilterNode node, CteContext context)
+        {
+            PlanNode childNode = node.getSource();
+            if (childNode instanceof CteConsumerNode) {
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java:148]: Lets reduce indentation by flipping if condition

```
if (!(childNode instanceof CteConsumerNode)) {
  return super.visitFilter(node, context);
}
...
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+                context.addCteConsumerInfo(cteName, producerColumns, ImmutableList.of(newPredicate));
+                return null;
+            }
+
+            return super.visitFilter(node, context);
+        }
+
+        public Void visitProject(ProjectNode node, CteContext context)
+        {
+            if (isCteConsumerFilterRestrict(node)) {
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java:163]: Same here to reduce indentation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+            RowExpression predicate = constant(true, BOOLEAN);
+            context.addCteConsumerInfo(cteName, producerColumns, ImmutableList.of(predicate));
+            return null;
+        }
+
+        @Override
+        public Void visitSequence(SequenceNode node, CteContext context)
+        {
+            List<PlanNode> cteProducers = node.getCteProducers();
+            for (int i = cteProducers.size() - 1; i >= 0; i--) {
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java:207]: Do we need to iterate in reverse order? If not, we can use utils like `forEach`

**Review by pranjalssh — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java (hunk 4 above)
Comment (mlyublena) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java:207]: we do need it like that, ctes may define new ctes used in subsequent cte nodes, and at the moment we've chosen to store them in this order. We may in the future refactor the representation to store ctes from the inner-most to outer-most scope in the right order.
CC: @jaystarshot

Comment (jaystarshot) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java:208]: Yeah we need to refactor that logic for scheduling improvements too, currently the execution follows reverse topological order

**Pull request merged**